ext/BijectorsEnzymeCoreExt.jl

```diff
@@ -0,0 +1,184 @@
+
+function EnzymeRules.forward(
+    ::Const{typeof(find_alpha)},
+    ::Type{RT},
+    wt_y::Union{Const,Duplicated,BatchDuplicated},
+    wt_u_hat::Union{Const,Duplicated,BatchDuplicated},
+    b::Union{Const,Duplicated,BatchDuplicated},
+) where {RT<:Union{Const,Duplicated,DuplicatedNoNeed,BatchDuplicated,BatchDuplicatedNoNeed}}
+    # Check that the types of the activities are consistent
+    if !(
+        RT <: Union{Const,Duplicated,DuplicatedNoNeed} &&
+        wt_y isa Union{Const,Duplicated} &&
+        wt_u_hat isa Union{Const,Duplicated} &&
+        b isa Union{Const,Duplicated}
+    ) && !(
+        RT <: Union{Const,BatchDuplicated,BatchDuplicatedNoNeed} &&
+        wt_y isa Union{Const,BatchDuplicated} &&
+        wt_u_hat isa Union{Const,BatchDuplicated} &&
+        b isa Union{Const,BatchDuplicated}
+    )
+        throw(ArgumentError("inconsistent activities"))
+    end
+
+    # Compute primal value
+    Ω = find_alpha(wt_y.val, wt_u_hat.val, b.val)
+
+    # Early exit if no derivatives are requested
+    if RT <: Const
+        return Ω
+    end
+
+    Ω̇ = if wt_y isa Const && wt_u_hat isa Const && b isa Const
+        # Trivial case: All partial derivatives are 0
+        zero(Ω)
+    else
+        # In all other cases we have to compute the partial derivatives
+        ∂Ω_∂wt_y, ∂Ω_∂wt_u_hat, ∂Ω_∂b = ∂find_alpha(Ω, wt_y, wt_u_hat, b)
+        _muladd_partial(
+            ∂Ω_∂wt_y,
+            wt_y,
+            _muladd_partial(∂Ω_∂wt_u_hat, wt_u_hat, _muladd_partial(∂Ω_∂b, b, nothing)),
+        )
+    end
+
+    if RT <: Duplicated
+        @assert Ω̇ isa Real
+        return Duplicated(Ω, Ω̇)
+    elseif RT <: DuplicatedNoNeed
+        @assert Ω̇ isa Real
+        return Ω̇
+    elseif RT <: BatchDuplicated
+        @assert Ω̇ isa Tuple{Vararg{Real}}
+        return BatchDuplicated(Ω, Ω̇)
+    else
+        @assert RT <: BatchDuplicatedNoNeed
+        @assert Ω̇ isa Tuple{Vararg{Real}}
+        return Ω̇
+    end
+end
+
+function EnzymeRules.augmented_primal(
+    config::EnzymeRules.Config,
+    ::Const{typeof(find_alpha)},
+    ::Type{RT},
+    wt_y::Union{Const,Active},
+    wt_u_hat::Union{Const,Active},
+    b::Union{Const,Active},
+) where {RT<:Union{Const,Active}}
+    # Only compute the the original return value if it is actually needed
+    Ω =
+        if EnzymeRules.needs_primal(config) ||
+            EnzymeRules.needs_shadow(config) ||
+            (RT <: Active && !(wt_y isa Const && wt_u_hat isa Const && b isa Const))
+            find_alpha(wt_y.val, wt_u_hat.val, b.val)
+        else
+            nothing
+        end
+
+    tape = if RT <: Const || (wt_y isa Const && wt_u_hat isa Const && b isa Const)
+        # Trivial case: No differentiation or all partial derivatives are 0
+        # Thus no tape is needed
+        nothing
+    else
+        # Store the partial derivatives
+        ∂find_alpha(Ω, wt_y, wt_u_hat, b)
```

Comment:
I know it's legal to run ∂find_alpha here, but why not move that to the reverse pass, and potentially cache Ω (or even recompute it if desirable)

Reply:
this then lets you simplify the reverse rule too since you can make dfind_alpha take ΔΩ_val as an argument and just return it

Reply:
What would be the advantage of moving it? In ChainRules, usually such computations are intentionally performed outside of the pullback such that they do not have to be recomputed when the pullback is reevaluated (eg this is also done by the `@scalar_rule` macro).

Reply:
Less data needs to be cached from fwd to rev. We also know the reverse pass is executed only once

Reply:
> We also know the reverse pass is executed only once

That means the motivation for the common ChainRules design (precomputing as much as possible outside of the pullback) does not exist in Enzyme?

Reply:
Correct

Reply:
Is this mentioned in the docs somewhere? I think it would be very useful to explain this, in particular for defining custom rules, I guess people familiar with ChainRules might be biased towards precomputing and caching stuff.

Reply:
I don’t think so, since I at least was unaware of the chainrules convention.

feel free to open a PR